core-bundle/src/Command/DeferredImagesCommand.php

```diff
@@ -0,0 +1,243 @@
+        $this->storage = $storage;
+        $this->terminalWidth = (new Terminal())->getWidth();
+
+        parent::__construct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure(): void
+    {
+        $this
+            ->setName('contao:deferred-images')
+            ->setDefinition([
+                new InputOption('time-limit', 'l', InputArgument::OPTIONAL, 'Time limit in seconds', '0'),
+                new InputOption('concurrent', 'c', InputArgument::OPTIONAL, 'Run multiple processes concurrently', '1'),
+                new InputOption('throttle', 't', InputArgument::OPTIONAL, 'Pause between resizes to limit cpu utilization, 0.1 relates to 10% cpu usage', '1'),
+            ])
+            ->setDescription('Resizes deferred images that have not been processed yet.')
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        if (!$this->resizer instanceof DeferredResizerInterface) {
+            throw new \RuntimeException('Deferred resizer not available');
+        }
+
+        $timeLimit = (float) $input->getOption('time-limit');
+        $throttle = (float) $input->getOption('throttle');
+        $concurrent = (string) $input->getOption('concurrent');
+
+        if ($timeLimit < 0) {
+            throw new InvalidArgumentException(sprintf('Time-limit value "%s" is invalid.', $timeLimit));
+        }
+
+        if ($throttle < 0.001 || $throttle > 1) {
+            throw new InvalidArgumentException(sprintf('Throttle value "%s" is invalid.', $timeLimit));
+        }
+
+        if ((string) (int) $concurrent !== $concurrent || (int) $concurrent < 1) {
+            throw new InvalidArgumentException(sprintf('Concurrent value "%s" is invalid.', $concurrent));
+        }
+
+        if ((int) $concurrent > 1) {
+            return $this->executeConcurrent((int) $concurrent, $input, $output);
+        }
+
+        $startTime = microtime(true);
+
+        $maxDuration = 0;
+        $lastDuration = 0;
+
+        foreach ($this->storage->listPaths() as $path) {
+            $sleep = (1 - $throttle) / $throttle * $lastDuration;
+
+            if ($timeLimit && microtime(true) - $startTime + $maxDuration + $sleep > $timeLimit) {
+                $output->writeln("\n".'Time limit of '.$timeLimit.' seconds reached.');
+
+                return 0;
+            }
+
+            usleep((int) ($sleep * 1000000));
+
+            $maxDuration = max($maxDuration, $lastDuration = $this->resizeImage($path, $output));
+        }
+
+        $output->writeln("\n".'All images resized.');
+
+        return 0;
+    }
+
+    private function resizeImage(string $path, OutputInterface $output): float
+    {
+        $startTime = microtime(true);
+
+        if ($this->filesystem->exists($path)) {
+            return 0;
+        }
+
+        $output->write(str_pad($path, $this->terminalWidth - 14, '.').' ');
```

Comment:
Can't we re-use some Symfony logic here? There are loads of helpers, maybe one of them might fit our needs? Table? Section?

Reply:
I didn’t find one for this kind of output. And I wanted the image path to be printed before the resize, so that if the process hangs the output shows which image is the problem.